### PR TITLE
vdk-jupyter: remove cell outputs before deployment

### PR DESCRIPTION
**Pull request by duyguHsnHsn**
What: added fuctionality to remove cell outputs before deployment 

Why: Currently, we do not remove cell outputs when deploying a notebook job which makes it really hard for version control and populates the deployed notebook with unneeded information.

Testing: pytest tests

Signed-off-by: Duygu Hasan [hduygu@vmware.com](mailto:hduygu@vmware.com)

### CODE REVIEW

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/handlers.py [hunk 1]
```diff
@@ -165,6 +166,8 @@ class CreateDeploymentHandler(APIHandler):
     def post(self):
         input_data = self.get_json_body()
         try:
+            job = NotebookJobDirectory(input_data[VdkOption.PATH.value])
+            job.remove_outputs_from_all()
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/handlers.py:170]: I don't think it's good idea to remove the outputs of the local jupyter notebook. 

That would be rather surprising to the user. He click deploys and everything gets wiped out . 

Instead only the deployed version should be without outputs.

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/jupyter_notebook.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def remove_outputs_from_notebook(notebook_path):
+    with notebook_path.open("r") as f:
+        content = json.load(f)
+
+    for cell in content["cells"]:
+        if cell["cell_type"] == "code":
+            cell["outputs"] = []
+            cell["execution_count"] = None
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/jupyter_notebook.py:15]: I am not fan of this hard-coding and manual stuff. 

This will be hard to maintain. If notebooks change their structure. 
This is generally error prone. How sure are you this is enough and everything you need to clean. 

We are manipulating cells in multiple places manually. Fortunately it's rather few places and code is not big. But it's a concern. 

Have you seen  https://nbconvert.readthedocs.io/en/latest/api/preprocessors.html#removing-cells-inputs-and-outputs

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/handlers.py (hunk 1 above)
Comment (duyguHsnHsn) [reply projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/handlers.py]: i have changed the implementation to work with a copy of the job

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/jupyter_notebook.py (hunk 2 above)
Comment (duyguHsnHsn) [reply projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/jupyter_notebook.py]: i have changed the implementation to work with it

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/jupyter_notebook.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2021-2023 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+import nbformat
+from nbconvert.preprocessors import ClearOutputPreprocessor
+
+
+class NotebookOutputCleaner:
```
Comment (DeltaMichael) [inline projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/jupyter_notebook.py:8]: This class can simply be a function in vdk_ui.py.

Comment (duyguHsnHsn) [reply projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/jupyter_notebook.py:8]: done

**Review by DeltaMichael — APPROVED** [review]

**Review by antoniivanov — CHANGES_REQUESTED** [review]
Looks good to me. Let's double check some corner cases and the tests one more time.

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/vdk_ui.py [hunk 4]
```diff
@@ -163,20 +165,38 @@ def create_deployment(name: str, team: str, path: str, reason: str):
...
+            cmd.create(
+                name=name,
+                team=team,
+                job_path=destination_path,
+                reason=reason,
+                vdk_version=None,
+                enabled=True,
+            )
+            return (
+                f"Request to deploy job with name {name} and team {team} is sent successfully!"
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/vdk_ui.py:192]: Why not reuse the full description like in the CLI :

                f"Request to deploy Data Job {name} using version {deployment.job_version} finished successfully.\n"
                f"It would take a few minutes for the Data Job to be deployed in the server.\n"
                f"If notified_on_job_deploy option in config.ini is configured then "
                f"notification will be sent on successful deploy or in case of an error.\n\n"

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/vdk_ui.py [hunk 5]
```diff
@@ -163,20 +165,37 @@ def create_deployment(name: str, team: str, path: str, reason: str):
         :param reason: the reason of deployment
         :return: output string of the operation
         """
+        # Make a copy of the directory at the same location
+        destination_path = os.path.join(path, os.path.basename(path))
+        shutil.copytree(path, destination_path)
```
[the remark this answers is not in the document]
Comment (duyguHsnHsn) [reply projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/vdk_ui.py]: added a check for it

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/tests/test_remove_cell_outputs.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    shutil.copytree(backup_dir, base_dir)
+    shutil.rmtree(backup_dir)
+
+
+@pytest.mark.parametrize("job_dir", ["ingest-job", "mixed-job"])
+def test_notebook_output_cleaner(job_dir):
+    base_dir = Path(__file__).parent / "jobs"
+    dir_path = base_dir / job_dir
+
+    for notebook_path in dir_path.glob("*.ipynb"):
```
[the remark this answers is not in the document]
Comment (duyguHsnHsn) [reply projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/tests/test_remove_cell_outputs.py:34]: pelase check the new tests i added

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/vdk_ui.py (hunk 4 above)
Comment (duyguHsnHsn) [reply projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/vdk_ui.py]: added the missing information

**Review by antoniivanov — APPROVED** [review]
